lib/puppet.rb

```diff
@@ -137,19 +137,23 @@ def self.run_mode
   #
   # @api public
   # @param args [Array<String>] the command line arguments to use for initialization
+  # @param require_config [Boolean] controls loading of Puppet configuration files
+  # @param global_settings [Boolean] controls loading of global context after settings initialization
   # @return [void]
-  def self.initialize_settings(args = [], require_config = true)
-    do_initialize_settings_for_run_mode(:user, args, require_config)
+  def self.initialize_settings(args = [], require_config = true, global_context = true)
+    do_initialize_settings_for_run_mode(:user, args, require_config, global_context)
   end
 
   # private helper method to provide the implementation details of initializing for a run mode,
   #  but allowing us to control where the deprecation warning is issued
-  def self.do_initialize_settings_for_run_mode(run_mode, args, require_config = true)
+  def self.do_initialize_settings_for_run_mode(run_mode, args, require_config = true, global_context = true)
     Puppet.settings.initialize_global_settings(args, require_config)
     run_mode = Puppet::Util::RunMode[run_mode]
     Puppet.settings.initialize_app_defaults(Puppet::Settings.app_defaults_for_run_mode(run_mode))
-    push_context_global(Puppet.base_context(Puppet.settings), "Initial context after settings initialization")
-    Puppet::Parser::Functions.reset
+    if global_context
+      push_context_global(Puppet.base_context(Puppet.settings), "Initial context after settings initialization")
+      Puppet::Parser::Functions.reset
```

Comment:
what's this do? does this need to be reset if we reset the "local" settings object?

Reply:
In the puppetserver PR, we call this method but set the global_context to false when are are generating the `local_settings` object in a thread. I think I called it "global_settings" in the code over there, so we will probably want to change it. `global_context` I think was more accurate, but it was a last minute change to this PR.

Reply:
Can we maybe get a verb or at least a question mark in the name of that var? As it is, it seems like you're expected to pass a context object.

Reply:
Sure, `push_settings_to_global_context` or something like that would work...I'm not sure if I think a question mark is a good idea, I feel like that's relatively uncommon for parameter names...but I could be down with it!

Reply:
Yeah I'm not sure either. I guess it's usually methods in Ruby.

Reply:
`update_global_context` perhaps?